Kernel/Process.cpp

```diff
@@ -1803,6 +1804,34 @@ int Process::sys$mkdir(const char* pathname, mode_t mode)
     return VFS::the().mkdir(StringView(pathname, pathname_length), mode & ~umask(), current_directory());
 }
 
+int Process::sys$realpath(const char* pathname, char* buffer, ssize_t size)
+{
+    if (size < 0)
+        return -EINVAL;
+    if (!validate_read_str(pathname))
+        return -EFAULT;
+
+    size_t pathname_length = strlen(pathname);
+    if (pathname_length == 0)
+        return -EINVAL;
+    if (pathname_length >= 255)
+        return -ENAMETOOLONG;
+
+    auto custody_or_error = VFS::the().resolve_path(pathname, current_directory());
+    if (custody_or_error.is_error())
+        return custody_or_error.error();
+    auto& custody = custody_or_error.value();
+
+    FileSystemPath canonical_path(custody->absolute_path());
```

Comment:
Why use `FileSystemPath` here instead of just the string? `resolve_path()` should already deal with all the `..`s

Reply:
If I skip `FileSystemPath` here, `..`s aren't dealt with. 

For example, `realpath ../anon` returns `/home/anon/../anon`.

Reply:
That would be a bug in `VFS::resolve_path()` then. `Custody::absolute_path()` should return an absolute path, but it doesn't do much itself. Whoever constructs the `Custody` — and that's `VFS::resolve_path()` — is responsible for constructing it properly.

So, I would say the right thing to do is to make `VFS::resolve_path()` pop a custody from the chain when it encounters a `..`. @awesomekling can correct me of course :smile:

Reply:
Oh yeah, that sounds like a bug in `VFS::resolve_path()`. I think @bugaevc's assessment is correct: it should pop one custody off from the chain for each `..` encountered.

Reply:
I've added FIXMEs so this can be dealt with in another PR.